File: 22-Generate-Parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;

        generate(ans, "", n, n);
        return ans;
    }

private:
    void generate(vector<string> &res, string curr, int open, int close) {
        if (open == 0 && close == 0) {
            res.push_back(curr);
            return;
        }

        if (open > 0)
            generate(res, curr + '(', open - 1, close);
        
        if (close > open)
            generate(res, curr + ')', open, close - 1);
    }
};

File: 22-Generate-Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []

        def generate(curr: str) -> None:
            if len(curr) == n * 2:
                if isValid(curr):
                    res.append(curr)
                return

            generate(curr + "(")
            generate(curr + ")")

        def isValid(s: str) -> bool:
            balance = 0

            for c in s:
                if c == "(":
                    balance += 1
                elif c == ")":
                    balance -= 1
                if balance < 0:
                    return False

            return balance == 0

        generate("")
        return res
